microsoft/STL
# Test P1502R1 Standard Library Header Units

Works towards #60.

* `stl/inc/memory_resource`
  + Work around DevCom-1159869 "Standard Library Header Units: Bogus warning C4373 about virtual function overriding" by dropping `const` on these value parameters. This has been fixed in VS 2019 16.9 Preview 1.
* `tests/std/test.lst`
  + Add the new test.
* `tests/std/tests/P1502R1_standard_library_header_units`
  + `__init__.py`
    - Complicated stuff happening here.
  + `custom_format.py`
    - Build and run the test in the GitHub Python test harness. We iterate through a list of the `stl_headers`, building each header unit with `/exportHeader`, and recording the `/headerUnit` compiler option and object file path needed to consume it later. (This also contains code to handle `/BE`, which has to be ignored when building the header unit. `/BE` is not yet active.)
  + `custombuild.pl`
    - Build and run the test in the MSVC-internal Perl test harness. I haven't written code to handle `/BE` yet. This needs a new environment variable `STL_INCLUDE_DIR` (added in MSVC-internal `distrib.yaml`) which is the single directory where STL headers live.
    - Because the compiler's implementation of modules is quickly moving, and we have two different test harnesses, I'm doing something special here. The Perl test harness will compile with `/DMSVC_INTERNAL_TESTING`, allowing us to activate portions of the test that successfully compile and run with the latest development build, before those fixes ship in the public VS Preview. As the TRANSITION comments note, this is temporary - after things settle down, we'll be able to remove this macro.
  + `env.lst`
    - This test is special, so it has its own configurations. (For example, modules require `/Zc:preprocessor`.) The apparently-strange pattern of `RUNALL_CROSSLIST`s is imitating how the `usual_matrix` family works.
    - This test takes a long time to compile (building each header unit sequentially), so I'm currently testing only the 4 main combinations of release/debug and dynamic/static. `/analyze` works but is *much* slower; as it found no compiler bugs, I have disabled it for now. `/BE` is blocked by Microsoft-internal VSO-1232145 "EDG ICEs when consuming Standard Library Header Units".
  + `lit.local.cfg`
    - I carefully copy-paste-modified this file and aligned its whitespace.
  + `test.cpp`
    - Test the entire STL. :joy_cat:
    - The intent of this test is to import every STL header as a header unit, and then minimally exercise its compile-time and/or run-time functionality. This has found many compiler bugs that would block users; see #60 for the list.
    - This does *not* attempt to exercise every class or function in the STL.

Adding this test to GitHub and MSVC now will prevent further regressions as we continue to change the STL and @cdacamar continues to fix the compiler. When all of the major blocking compiler bugs are fixed, we'll be able to declare feature-completeness.

Microsoft-internal mirror: MSVC-PR-282732
